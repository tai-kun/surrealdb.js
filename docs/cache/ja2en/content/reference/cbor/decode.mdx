---
title: decode
---

## Synchronous Decoding

The `decode` function decodes a CBOR-formatted `Uint8Array` into a JavaScript value.

### Import

```ts
import { decode } from "@tai-kun/surrealdb/cbor";
```

### Syntax

```ts
function decode(input: Uint8Array, options?: DecodeOptions): unknown;
```

#### Arguments

##### `input`

The CBOR-formatted `Uint8Array` to decode into a JavaScript value.

##### `options`

Decoding options.

`mapType?: "Object" | "Map"`
:   This option can be used to choose whether to decode key-value objects as JavaScript plain objects or `Map` objects. Choose `"Object"` to decode as a plain object and `"Map"` to decode as a `Map` class instance. The default is `"Object"`. Plain object keys can only be strings or numbers (i.e., `string | number`), but if you decode with this option set to `"Map"`, the keys can be any decoded value.

`reviver?: Reviver | ReviverObject | (Reviver | ReviverObject)[]`
:   This option is used to transform CBOR simple data items or tagged data items into JavaScript values. Simple data items are limited to values from 0 to 19 and 32 to 255. The tag of a tagged data item is a `number` or `bigint` from 0 to 2^64-1. If the function returns the `CONTINUE` symbol, which can be obtained with `Symbol.for("@tai-kun/surrealdb/cbor/continue")`, subsequent functions will be called. If the value cannot be transformed, `SurrealTypeError` is thrown. 

`isSafeMapKey?: (key: unknown, map: Map<unknown, unknown>) => boolean`
:   This option is a function that validates whether a key of a `Map` object is safe. The function takes two arguments: `key` and `map`. `key` is the value of the key, and `map` is the `Map` object being validated. If this function returns `false`, `CborUnsafeMapKeyError` is thrown. The default considers `"__proto__"` or `"constructor"` as unsafe keys. Note that specifying this option will completely override the default validation. If you want to add your own validation to the default validation, you should also consider `"__proto__"` or `"constructor"` as unsafe keys.

`isSafeObjectKey?: (key: unknown, obj: Record<string | number, unknown>) => boolean`
:   This option is a function that validates whether a key of a plain object is safe. The function takes two arguments: `key` and `obj`. `key` is the value of the key, and `obj` is the plain object being validated. If this function returns `false`, `CborUnsafeMapKeyError` is thrown. The default considers `"__proto__"` or `"constructor"` as unsafe keys. Note that specifying this option will completely override the default validation. If you want to add your own validation to the default validation, you should also consider `"__proto__"` or `"constructor"` as unsafe keys.

`maxDepth?: number`
:   Specifies the maximum depth of the object. The depth is incremented by 1 for each object or array entered. The default value is `64`. Since recursion is not used for decoding, there is no upper limit to the depth, but it is recommended to keep it small.

#### Return Value

The JavaScript value decoded from the CBOR-formatted `Uint8Array`.

### Example

#### Decoding as a `Map` object

```ts
import { decode } from "@tai-kun/surrealdb/cbor";

const cbor = new Uint8Array([
  0xa1, // mt: 5, ai: 1
  // key
  0x00, // mt: 0, ai: 0
  // value
  0x00, // mt: 0, ai: 0
]);

const value = decode(bytes, { mapType: "Map" });

console.log(value); // Map(1) { [ 0 ] => 0 }
```

## Asynchronous Decoding

The `decodeStream` function decodes a CBOR-formatted `Uint8Array` into a JavaScript value.

### Import

```ts
import { decodeStream } from "@tai-kun/surrealdb/cbor";
```

### Syntax

```ts
function decodeStream(
  input: ReadableStream<Uint8Array>,
  options?: DecodeStreamOptions,
): unknown;
```

#### Arguments

##### `input`

A readable stream of `Uint8Array`. The Fetch API may provide this stream on the `body` property of the `Response` object in response to a request.

##### `options`

In addition to the options for synchronous decoding, the following options can be specified:

`signal?: AbortSignal`
:   An AbortSignal to abort the stream.

#### Return Value

A `StatefulPromise` that resolves with the JavaScript value.

### Example

#### Fetch API

```ts
import { decodeStream } from "@tai-kun/surrealdb/cbor";

const response = await fetch("http://localhost:8000/rpc", {
  method: "POST",
  headers: {
    Accept: "application/cbor",
    // ... more headers
  },
  // ... more properties
})

const result = await decodeStream(resp.body);
```
