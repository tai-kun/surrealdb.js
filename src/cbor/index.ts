export type * from "./decode-stream";
export { default as decodeStream } from "./decode-stream";

export type * from "./decode";
export { default as decode } from "./decode";

export {
  CONTINUE,
  type IsSafeMapKey,
  type IsSafeObjectKey,
  type Reviver,
  type ReviverObject,
  type SimpleItemReviver,
  type TaggedItemReviver,
} from "./decoder";

export type * from "./encode";
export { default as encode } from "./encode";

export {
  type AdditionalInfo,
  AI_EIGHT_BYTES,
  AI_FOUR_BYTES,
  AI_INDEFINITE_NUM_BYTES,
  AI_ONE_BYTE,
  AI_SIMPLE_FALSE,
  AI_SIMPLE_NULL,
  AI_SIMPLE_TRUE,
  AI_SIMPLE_UNDEFINED,
  AI_TWO_BYTES,
  type DataItem,
  type DataItemValue,
  HEADER_FALSE,
  HEADER_FLOAT_DOUBLE,
  HEADER_FLOAT_HALF,
  HEADER_NULL,
  HEADER_TRUE,
  HEADER_UNDEFINED,
  type MajorType,
  MT_ARRAY,
  MT_BYTE_STRING,
  MT_MAP,
  MT_NEGATIVE_INTEGER,
  MT_SIMPLE_FLOAT,
  MT_TAG,
  MT_UNSIGNED_INTEGER,
  MT_UTF8_STRING,
  Simple,
} from "./spec";

export type * from "./tagged";
export { default as Tagged } from "./tagged";

export type { ToCBOR } from "./traits";
export { canToCBOR } from "./traits";

export {
  type Replacer,
  write,
  writeBigInt,
  writeBoolean,
  writeByteString,
  writeEncodedUtf8String,
  writeHeader,
  writeNullable,
  writeNumber,
  writePayload,
  writeUtf8String,
} from "./write-utils";

export type * from "./writer"; // `Writer` も型だけでよし
