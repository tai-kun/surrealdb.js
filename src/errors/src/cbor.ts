import { SurrealError, type SurrealErrorOptions } from "./general";

/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cborerror)
 */
export class CborError extends SurrealError {
  static {
    this.prototype.name = "CborError";
  }
}

// https://datatracker.ietf.org/doc/html/rfc8949#name-well-formedness-errors-and-
/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cborwellformednesserror)
 */
export class CborWellFormednessError extends CborError {
  static {
    this.prototype.name = "CborWellFormednessError";
  }
}

// https://datatracker.ietf.org/doc/html/rfc8949#section-appendix.f-2.2
/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cbortoomuchdataerror)
 */
export class CborTooMuchDataError extends CborWellFormednessError {
  static {
    this.prototype.name = "CborTooMuchDataError";
  }

  constructor(options?: SurrealErrorOptions | undefined) {
    super("Unconsumed input bytes remain after decoding.", options);
  }
}

// https://datatracker.ietf.org/doc/html/rfc8949#section-appendix.f-2.4
/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cbortoolittledataerror)
 */
export class CborTooLittleDataError extends CborWellFormednessError {
  static {
    this.prototype.name = "CborTooLittleDataError";
  }

  constructor(options?: SurrealErrorOptions | undefined) {
    super(
      "Input data appears truncated or incomplete for CBOR decoding.",
      options,
    );
  }
}

// https://datatracker.ietf.org/doc/html/rfc8949#section-appendix.f-2.6
/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cborsyntaxerror)
 */
export class CborSyntaxError extends CborWellFormednessError {
  static {
    this.prototype.name = "CborSyntaxError";
  }
}

/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cbormaxdepthreachederror)
 */
export class CborMaxDepthReachedError extends CborError {
  static {
    this.prototype.name = "CborMaxDepthReachedError";
  }

  constructor(
    public maxDepth: number,
    options?: SurrealErrorOptions | undefined,
  ) {
    super(`Maximum depth of ${maxDepth} has been reached.`, options);
  }
}

/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cborunsafemapkeyerror)
 */
export class CborUnsafeMapKeyError extends CborError {
  static {
    this.prototype.name = "CborUnsafeMapKeyError";
  }

  constructor(
    public key: unknown,
    options?: SurrealErrorOptions | undefined,
  ) {
    super(
      `Invalid key for CBOR map: ${String(key)}. The key must be a valid CBOR`
        + " data type that is safe to use in a JavaScript map or object.",
      options,
    );
  }
}

export class CborDecodeStreamAbortFailedError extends CborError {
  static {
    this.prototype.name = "CborStreamAbortFailedError";
  }

  override cause: unknown[];

  constructor(
    errors: readonly unknown[],
    options?: Omit<SurrealErrorOptions, "cause"> | undefined,
  ) {
    super("Failed to abort decode-stream", options);
    this.cause = errors.slice();
  }
}

export class CborMemoryError extends CborError {
  static {
    this.prototype.name = "CborMemoryError";
  }
}

export class CborMemoryBlockError extends CborError {
  static {
    this.prototype.name = "CborMemoryBlockError";
  }
}

type MemoryAddress = string | number;

export class CborMemoryBlockConflictError extends CborMemoryBlockError {
  static {
    this.prototype.name = "CborMemoryBlockConflictError";
  }

  constructor(
    address: MemoryAddress,
    expectedSize: number,
    actualSize: number,
  ) {
    super(
      `Memory conflict at address ${address}: `
        + `Expected size ${expectedSize}, but found ${actualSize}.`,
    );
  }
}

export class CborMemoryBlockInUseError extends CborMemoryBlockError {
  static {
    this.prototype.name = "CborMemoryBlockInUseError";
  }
}

export class CborUndefinedMemoryBlockError extends CborMemoryBlockError {
  static {
    this.prototype.name = "CborUndefinedMemoryBlockError";
  }

  constructor(address: MemoryAddress) {
    super(`Memory block at address ${address} is not defined.`);
  }
}
