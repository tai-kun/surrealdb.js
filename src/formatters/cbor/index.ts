export type * from "./formatter";
export { default } from "./formatter";
