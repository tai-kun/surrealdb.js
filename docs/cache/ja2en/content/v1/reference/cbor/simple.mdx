---
title: Simple
slug: en/v1/reference/cbor/simple
---

import Badge from "~/components/en/Badge.astro";

`Simple` is a simple value identified by an integer value from 0 to 255.

### Import

```ts
import { Simple } from "@tai-kun/surrealdb/cbor";
```

### `.constructor()`

Creates a new `Simple` object.

```ts
new Simple(value: number);
```

#### Arguments

##### `value`

An integer value from 0 to 255.

#### Return Value

When called via `new`, `Simple` returns its instance.

### `.value` <Badge variant="instance" /><Badge variant="property" />

The integer value from 0 to 255 that identifies the `Simple` object.

### Notes

Refer to the following materials:

- [https://datatracker.ietf.org/doc/html/rfc8949#name-simple-values](https://datatracker.ietf.org/doc/html/rfc8949#name-simple-values)
- [https://www.iana.org/assignments/cbor-simple-values/cbor-simple-values.xhtml](https://www.iana.org/assignments/cbor-simple-values/cbor-simple-values.xhtml)
