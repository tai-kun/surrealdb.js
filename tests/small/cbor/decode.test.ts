import { decode } from "@tai-kun/surrealdb/cbor";
import {
  CborMaxDepthReachedError,
  CborUnsafeMapKeyError,
} from "@tai-kun/surrealdb/errors";
import { describe, expect, test } from "vitest";

describe("map", () => {
  test("プレーンオブジェクトの代わりに Map オブジェクト", () => {
    const bytes = new Uint8Array([
      0xa1, // mt: 5, ai: 1
      // key
      0x81, // mt: 4, ai: 1
      0x00, // mt: 0, ai: 0
      // value
      0x00, // mt: 0, ai: 0
    ]);

    expect(decode(bytes, { mapType: "Map" })).toStrictEqual(
      new Map([
        [[0], 0],
      ]),
    );
  });
});

describe("isSafeMapKey", () => {
  test("__proto__", () => {
    const bytes = new Uint8Array([
      0xa1, // mt: 5, ai: 1
      // key
      0b011_01001, // mt: 3, ai: 9
      0x5f, // _
      0x5f, // _
      0x70, // p
      0x72, // r
      0x6f, // o
      0x74, // t
      0x6f, // o
      0x5f, // _
      0x5f, // _
      // value
      0x00, // mt: 0, ai: 0
    ]);

    const run = () => {
      decode(bytes, { mapType: "Map" });
    };
    expect(run).toThrowErrorMatchingSnapshot();
  });

  test("constructor", () => {
    const bytes = new Uint8Array([
      0xa1, // mt: 5, ai: 1
      // key
      0b011_01011, // mt: 3, ai: 11
      0x63, // c
      0x6f, // o
      0x6e, // n
      0x73, // s
      0x74, // t
      0x72, // r
      0x75, // u
      0x63, // c
      0x74, // t
      0x6f, // o
      0x72, // r
      // value
      0x00, // mt: 0, ai: 0
    ]);

    const run = () => {
      decode(bytes, { mapType: "Map" });
    };
    expect(run).toThrowErrorMatchingSnapshot();
  });

  test("カスタマイズ", () => {
    const bytes = new Uint8Array([
      0xa1, // mt: 5, ai: 1
      // key
      0b000_00001, // mt: 0, ai: 1
      // value
      0x02, // mt: 0, ai: 2
    ]);

    const run = () => {
      decode(bytes, {
        mapType: "Map",
        isSafeMapKey: k => k !== 1,
      });
    };
    expect(run).toThrowErrorMatchingSnapshot();
  });

  test("カスタマイズはオブジェクトに影響しない", () => {
    const bytes = new Uint8Array([
      0xa1, // mt: 5, ai: 1
      // key
      0b011_00001, // mt: 3, ai: 1
      0x00, // mt: 0, ai: 1
      // value
      0x02, // mt: 0, ai: 2
    ]);

    const run = () => {
      decode(bytes, {
        // mapType: "Map",
        isSafeMapKey: k => k === 1,
      });
    };
    expect(run).not.toThrowError(CborUnsafeMapKeyError);
  });
});

describe("isSafeObjectKey", () => {
  test("__proto__", () => {
    const bytes = new Uint8Array([
      0xa1, // mt: 5, ai: 1
      // key
      0b011_01001, // mt: 3, ai: 9
      0x5f, // _
      0x5f, // _
      0x70, // p
      0x72, // r
      0x6f, // o
      0x74, // t
      0x6f, // o
      0x5f, // _
      0x5f, // _
      // value
      0x00, // mt: 0, ai: 0
    ]);

    const run = () => {
      decode(bytes);
    };
    expect(run).toThrowErrorMatchingSnapshot();
  });

  test("constructor", () => {
    const bytes = new Uint8Array([
      0xa1, // mt: 5, ai: 1
      // key
      0b011_01011, // mt: 3, ai: 11
      0x63, // c
      0x6f, // o
      0x6e, // n
      0x73, // s
      0x74, // t
      0x72, // r
      0x75, // u
      0x63, // c
      0x74, // t
      0x6f, // o
      0x72, // r
      // value
      0x00, // mt: 0, ai: 0
    ]);

    const run = () => {
      decode(bytes);
    };
    expect(run).toThrowErrorMatchingSnapshot();
  });

  test("カスタマイズ", () => {
    const bytes = new Uint8Array([
      0xa1, // mt: 5, ai: 1
      // key
      0b000_00001, // mt: 0, ai: 1
      // value
      0x02, // mt: 0, ai: 2
    ]);

    const run = () => {
      decode(bytes, {
        isSafeObjectKey: k => k !== 1,
      });
    };
    expect(run).toThrowErrorMatchingSnapshot();
  });

  test("カスタマイズは Map に影響しない", () => {
    const bytes = new Uint8Array([
      0xa1, // mt: 5, ai: 1
      // key
      0b011_00001, // mt: 3, ai: 1
      0x00, // mt: 0, ai: 1
      // value
      0x02, // mt: 0, ai: 2
    ]);

    const run = () => {
      decode(bytes, {
        mapType: "Map",
        isSafeObjectKey: k => k === 1,
      });
    };
    expect(run).not.toThrowError(CborUnsafeMapKeyError);
  });
});

describe("maxDepth", () => {
  test("最大深さと同じ深さなら大丈夫", () => {
    // [[]]
    const bytes = new Uint8Array([
      0x81, // mt: 4, ai: 1
      0x80, // mt: 4, ai: 0
    ]);

    const run = () => {
      decode(
        bytes,
        { maxDepth: 2 },
      );
    };
    expect(run).not.toThrowError(CborMaxDepthReachedError);
  });

  test("最大深さを超えるオブジェクトでエラー", () => {
    // [[[]]]
    const bytes = new Uint8Array([
      0x81, // mt: 4, ai: 1
      0x81, // mt: 4, ai: 1
      0x80, // mt: 4, ai: 0
    ]);

    const run = () => {
      const a = decode(
        bytes,
        { maxDepth: 2 },
      );
      console.dir(a, { depth: null });
    };
    expect(run).toThrowErrorMatchingSnapshot();
  });
});

describe("Well-Formedness Errors", () => {
  // 現在未使用
  // describe("Too much data", () => {});

  describe("Too little data", () => {
    test("文字数が足りない", () => {
      const bytes = new Uint8Array([
        0b011_00011, // mt: 3, ai: 3 <- 3 bytes
        0b010_01000, // H <- 1/3
        0b011_00101, // e <- 2/3
        // 0b011_11001, // y <- 3/3 (LOST)
      ]);

      expect(() => decode(bytes)).toThrowErrorMatchingSnapshot();
    });
  });

  describe("Syntax error", () => {
    for (const mt of [0, 1, 2, 3, 4, 5, 6, 7]) {
      for (const ai of [28, 29, 30]) {
        // dprint-ignore
        const bin =
            mt.toString(2).padStart(3, "0")
          + ai.toString(2).padStart(5, "0");

        test(`0b${bin} (mt=${mt},ai=${ai}) は予約済み`, () => {
          const bytes = new Uint8Array([
            Number.parseInt(bin, 2),
          ]);

          expect(() => decode(bytes)).toThrowErrorMatchingSnapshot();
        });
      }
    }

    test("正の整数は不定長バイトをとれない", () => {
      const bytes = new Uint8Array([
        0b000_11111,
      ]);

      expect(() => decode(bytes)).toThrowErrorMatchingSnapshot();
    });

    test("負の整数は不定長バイトをとれない", () => {
      const bytes = new Uint8Array([
        0b001_11111,
      ]);

      expect(() => decode(bytes)).toThrowErrorMatchingSnapshot();
    });

    test("タグは不定長バイトをとれない", () => {
      const bytes = new Uint8Array([
        0b110_11111,
      ]);

      expect(() => decode(bytes)).toThrowErrorMatchingSnapshot();
    });

    test("不定長バイトの開始なく BREAK できない", () => {
      const bytes = new Uint8Array([
        0xff,
      ]);

      expect(() => decode(bytes)).toThrowErrorMatchingSnapshot();
    });

    test("不定長のバイト文字列は全データアイテムのメジャータイプが 4", () => {
      const bytes = new Uint8Array([
        0b010_11111, // mt: 2, ai: 31
        0b010_00011, // mt: 3, ai: 3
        0b010_00000,
        0b010_00000,
        0b010_00000,
        0b000_00000, // mt: 0, ai: 0 <- ERROR
        0b111_11111, // BREAK
      ]);

      expect(() => decode(bytes)).toThrowErrorMatchingSnapshot();
    });

    test("不定長の UTF-8 文字列は全データアイテムのメジャータイプが 3", () => {
      const bytes = new Uint8Array([
        0b011_11111, // mt: 3, ai: 31
        0b011_00011, // mt: 3, ai: 3
        0b010_01000, // H <- OK (ASCII)
        0b011_00101, // e <- OK (ASCII)
        0b011_11001, // y <- OK (ASCII)
        0b000_00000, // mt: 0, ai: 0 <- ERROR
        0b111_11111, // BREAK
      ]);

      expect(() => decode(bytes)).toThrowErrorMatchingSnapshot();
    });

    test("不定長のバイト文字列に不定長のバイト文字列だめ", () => {
      const bytes = new Uint8Array([
        0b010_11111, // mt: 2, ai: 31
        0b010_00011, // mt: 3, ai: 3
        0b010_00000,
        0b010_00000,
        0b010_00000,
        ...[
          0b010_11111, // mt: 2, ai: 31 <- ERROR
          0b010_00011, // mt: 3, ai: 3
          0b010_00000,
          0b010_00000,
          0b010_00000,
          0b111_11111, // BREAK
        ],
        0b111_11111, // BREAK
      ]);

      expect(() => decode(bytes)).toThrowErrorMatchingSnapshot();
    });

    test("不定長 UTF-8 文字列に不定長 UTF-8 文字列だめ", () => {
      const bytes = new Uint8Array([
        0b011_11111, // mt: 3, ai: 31
        0b011_00011, // mt: 3, ai: 3
        0b010_01000, // H <- OK (ASCII)
        0b011_00101, // e <- OK (ASCII)
        0b011_11001, // y <- OK (ASCII)
        ...[
          0b011_11111, // mt: 3, ai: 31 <- ERROR
          0b011_00011, // mt: 3, ai: 3
          0b010_01000, // H
          0b011_00101, // e
          0b011_11001, // y
          0b111_11111, // BREAK
        ],
        0b111_11111, // BREAK
      ]);

      expect(() => decode(bytes)).toThrowErrorMatchingSnapshot();
    });
  });
});
