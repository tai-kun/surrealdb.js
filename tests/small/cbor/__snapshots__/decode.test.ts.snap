// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Well-Formedness Errors > Syntax error > 0b00011100 (mt=0,ai=28) は予約済み 1`] = `
{
  "message": "The additional info 28 is reserved for future additions to the CBOR format.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 0b00011101 (mt=0,ai=29) は予約済み 1`] = `
{
  "message": "The additional info 29 is reserved for future additions to the CBOR format.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 0b00011110 (mt=0,ai=30) は予約済み 1`] = `
{
  "message": "The additional info 30 is reserved for future additions to the CBOR format.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 0b00111100 (mt=1,ai=28) は予約済み 1`] = `
{
  "message": "The additional info 28 is reserved for future additions to the CBOR format.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 0b00111101 (mt=1,ai=29) は予約済み 1`] = `
{
  "message": "The additional info 29 is reserved for future additions to the CBOR format.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 0b00111110 (mt=1,ai=30) は予約済み 1`] = `
{
  "message": "The additional info 30 is reserved for future additions to the CBOR format.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 0b01011100 (mt=2,ai=28) は予約済み 1`] = `
{
  "message": "The additional info 28 is reserved for future additions to the CBOR format.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 0b01011101 (mt=2,ai=29) は予約済み 1`] = `
{
  "message": "The additional info 29 is reserved for future additions to the CBOR format.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 0b01011110 (mt=2,ai=30) は予約済み 1`] = `
{
  "message": "The additional info 30 is reserved for future additions to the CBOR format.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 0b01111100 (mt=3,ai=28) は予約済み 1`] = `
{
  "message": "The additional info 28 is reserved for future additions to the CBOR format.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 0b01111101 (mt=3,ai=29) は予約済み 1`] = `
{
  "message": "The additional info 29 is reserved for future additions to the CBOR format.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 0b01111110 (mt=3,ai=30) は予約済み 1`] = `
{
  "message": "The additional info 30 is reserved for future additions to the CBOR format.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 0b10011100 (mt=4,ai=28) は予約済み 1`] = `
{
  "message": "The additional info 28 is reserved for future additions to the CBOR format.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 0b10011101 (mt=4,ai=29) は予約済み 1`] = `
{
  "message": "The additional info 29 is reserved for future additions to the CBOR format.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 0b10011110 (mt=4,ai=30) は予約済み 1`] = `
{
  "message": "The additional info 30 is reserved for future additions to the CBOR format.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 0b10111100 (mt=5,ai=28) は予約済み 1`] = `
{
  "message": "The additional info 28 is reserved for future additions to the CBOR format.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 0b10111101 (mt=5,ai=29) は予約済み 1`] = `
{
  "message": "The additional info 29 is reserved for future additions to the CBOR format.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 0b10111110 (mt=5,ai=30) は予約済み 1`] = `
{
  "message": "The additional info 30 is reserved for future additions to the CBOR format.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 0b11011100 (mt=6,ai=28) は予約済み 1`] = `
{
  "message": "The additional info 28 is reserved for future additions to the CBOR format.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 0b11011101 (mt=6,ai=29) は予約済み 1`] = `
{
  "message": "The additional info 29 is reserved for future additions to the CBOR format.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 0b11011110 (mt=6,ai=30) は予約済み 1`] = `
{
  "message": "The additional info 30 is reserved for future additions to the CBOR format.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 0b11111100 (mt=7,ai=28) は予約済み 1`] = `
{
  "message": "The additional info 28 is reserved for future additions to the CBOR format.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 0b11111101 (mt=7,ai=29) は予約済み 1`] = `
{
  "message": "The additional info 29 is reserved for future additions to the CBOR format.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 0b11111110 (mt=7,ai=30) は予約済み 1`] = `
{
  "message": "The additional info 30 is reserved for future additions to the CBOR format.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > タグは不定長バイトをとれない 1`] = `
{
  "message": "The data item \`tagged item\` has indefinite-length bytes.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 不定長 UTF-8 文字列に不定長 UTF-8 文字列だめ 1`] = `
{
  "message": "The payload of the indefinite-length data item \`utf8 string\` contains the same type.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 不定長の UTF-8 文字列は全データアイテムのメジャータイプが 3 1`] = `
{
  "message": "The payload of the data item \`utf8 string\` contains a mismatched data item \`unsigned integer\`.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 不定長のバイト文字列に不定長のバイト文字列だめ 1`] = `
{
  "message": "The payload of the indefinite-length data item \`byte string\` contains the same type.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 不定長のバイト文字列は全データアイテムのメジャータイプが 4 1`] = `
{
  "message": "The payload of the data item \`byte string\` contains a mismatched data item \`unsigned integer\`.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 不定長バイトの開始なく BREAK できない 1`] = `
{
  "message": "The "break" stop code was encountered outside of a indefinite-length data item.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 正の整数は不定長バイトをとれない 1`] = `
{
  "message": "The data item \`unsigned integer\` has indefinite-length bytes.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Syntax error > 負の整数は不定長バイトをとれない 1`] = `
{
  "message": "The data item \`negative integer\` has indefinite-length bytes.",
  "name": "CborSyntaxError",
}
`;

exports[`Well-Formedness Errors > Too little data > 文字数が足りない 1`] = `
{
  "cause": {
    "fragment": {
      "$": 0,
      "data": Uint8Array [
        72,
        101,
        0,
      ],
      "mt": 3,
      "offset": 2,
      "size": 3,
    },
    "parents": [
      {
        "$": 0,
        "out": [Function],
        "put": [Function],
        "val": undefined,
      },
    ],
  },
  "message": "Input data appears truncated or incomplete for CBOR decoding.",
  "name": "CborTooLittleDataError",
}
`;

exports[`isSafeMapKey > __proto__ 1`] = `
{
  "key": "__proto__",
  "message": "Invalid key for CBOR map: __proto__. The key must be a valid CBOR data type that is safe to use in a JavaScript map or object.",
  "name": "CborUnsafeMapKeyError",
}
`;

exports[`isSafeMapKey > constructor 1`] = `
{
  "key": "constructor",
  "message": "Invalid key for CBOR map: constructor. The key must be a valid CBOR data type that is safe to use in a JavaScript map or object.",
  "name": "CborUnsafeMapKeyError",
}
`;

exports[`isSafeMapKey > カスタマイズ 1`] = `
{
  "key": 1,
  "message": "Invalid key for CBOR map: 1. The key must be a valid CBOR data type that is safe to use in a JavaScript map or object.",
  "name": "CborUnsafeMapKeyError",
}
`;

exports[`isSafeObjectKey > __proto__ 1`] = `
{
  "key": "__proto__",
  "message": "Invalid key for CBOR map: __proto__. The key must be a valid CBOR data type that is safe to use in a JavaScript map or object.",
  "name": "CborUnsafeMapKeyError",
}
`;

exports[`isSafeObjectKey > constructor 1`] = `
{
  "key": "constructor",
  "message": "Invalid key for CBOR map: constructor. The key must be a valid CBOR data type that is safe to use in a JavaScript map or object.",
  "name": "CborUnsafeMapKeyError",
}
`;

exports[`isSafeObjectKey > カスタマイズ 1`] = `
{
  "key": 1,
  "message": "Invalid key for CBOR map: 1. The key must be a valid CBOR data type that is safe to use in a JavaScript map or object.",
  "name": "CborUnsafeMapKeyError",
}
`;

exports[`maxDepth > 最大深さを超えるオブジェクトでエラー 1`] = `
{
  "maxDepth": 2,
  "message": "Maximum depth of 2 has been reached.",
  "name": "CborMaxDepthReachedError",
}
`;
